gh-83854: Map `errno==ETIME` to `TimeoutError`

ZackerySpytz opened the pull request:
<!-- issue-number: [bpo-39673](https://bugs.python.org/issue39673) -->
https://bugs.python.org/issue39673
<!-- /issue-number -->


<!-- gh-issue-number: gh-83854 -->
* Issue: gh-83854
<!-- /gh-issue-number -->

ZackerySpytz committed:
bpo-39673: Map errno==ETIME to TimeoutError

YoSTEALTH reviewed (changes requested):
Doc/library/exceptions.rst needs to be updated as well

ZackerySpytz committed:
Update Doc/library/exceptions.rst

YoSTEALTH reviewed (approved):
looks good.

ericvsmith reviewed (commented):
Is it possible to add a test for this? I realize it might be difficult, so it's not a show stopper, but it would be good to test newly added code. There are a few ETIMEDOUT tests, so maybe they could be used as a guide?

ericvsmith commented on Misc/NEWS.d/next/Library/2020-05-19-23-16-53.bpo-39673.ppfScY.rst:
```diff
@@ -0,0 +1 @@
+Map ``errno==ETIME`` to :exc:`TimeoutError`.
```
Could you include in here that it's an OSError whose behavior is being changed? It's relevant that it's an OSError with errno==ETIME that's becoming a TimeoutError. Maybe just add "OSError with" after "Map"?

ericvsmith replied:
Since it's trivial enough, I made this edit myself.

ericvsmith committed:
Update 2020-05-19-23-16-53.bpo-39673.ppfScY.rst

Updated blurb to reflect that it's an OSError that's being mapped to a TimeoutError.

ericvsmith commented on Lib/test/test_exception_hierarchy.py:
```diff
@@ -65,7 +65,7 @@ def test_select_error(self):
         +-- PermissionError                                     EACCES, EPERM
         +-- ProcessLookupError                                          ESRCH
-        +-- TimeoutError                                            ETIMEDOUT
+        +-- TimeoutError                                     ETIME, ETIMEDOUT
```
I guess this test is as good as we're going to get.

ZackerySpytz replied:
To be honest, I'm having second thoughts w.r.t. this PR (especially after the discussion
on python-dev). I don't know a lot about ETIME. I'm not sure if I should have submitted this PR.

YoSTEALTH replied:
@ZackerySpytz No need to worry, its only being added to future version of python. With more and more async style coding being adapted ``ETIME`` will be handy to have as TimeoutError. It should have been adapted a long time ago but i suppose no one wanted to bother submitting a PR for it.

erlend-aasland committed:
Merge branch 'main' into bpo-39673-ETIME-TimeoutError
